File: web/kernel.js
/* The Recall Ops kernel, in the browser.
 *
 * A port of src/kernel.py (and fleet_view from src/views.py) for the static
 * live demo, where there is no Python server. It interprets the same
 * ontology/ontology.json. tests/test_parity.py runs the same scenarios through
 * both kernels and fails if they disagree on a single edit or error message,
 * so the two can't drift apart silently.
 *
 * One deliberate difference: the Python kernel refuses to start on a damaged
 * edit log. Here the log is the visitor's own browser storage, so damaged
 * entries are skipped with a warning instead of breaking the demo.
 */
(function (root) {
  "use strict";

  class ActionError extends Error {
    constructor(message, field = null) {
      super(message);
      this.name = "ActionError";
      this.field = field;
    }
  }

  const ACTIVE = new Set(["open", "scheduled"]);
  const CLOSED = new Set(["completed", "dismissed"]);
  const EFFECT_KEYS = ["objectType", "primaryKey", "property"];

  const isObject = v => v !== null && typeof v === "object" && !Array.isArray(v);
  const isString = v => typeof v === "string";
  const has = (obj, key) => isObject(obj) && Object.prototype.hasOwnProperty.call(obj, key);
  const own = (obj, key) => (has(obj, key) ? obj[key] : undefined);
  const codePoints = s => [...s].length; // Python's len() counts code points, not UTF-16 units

  function nowIso() {
    return new Date().toISOString().replace(/\.\d{3}Z$/, "+00:00");
  }

  // Exactly YYYY-MM-DD and a real calendar date, as in the Python kernel.
  function parseDate(raw) {
    if (!/^[0-9]{4}-[0-9]{2}-[0-9]{2}$/.test(raw)) return null;
    const [y, m, d] = raw.split("-").map(Number);
    const probe = new Date(0);
    probe.setUTCFullYear(y, m - 1, d);
    const real = y >= 1 && probe.getUTCFullYear() === y && probe.getUTCMonth() === m - 1 && probe.getUTCDate() === d;
    return real ? raw : null;
  }

  function isEdit(edit) {
    return isObject(edit) && isString(edit.id) && Array.isArray(edit.effects)
      && edit.effects.every(e => isObject(e) && has(e, "after") && EFFECT_KEYS.every(k => isString(e[k])));
  }

  function applyEdit(objects, edit, orphaned) {
    for (const effect of edit.effects) {
      const target = own(own(objects, effect.objectType), effect.primaryKey);
      if (target === undefined) {
        orphaned.push({ edit: edit.id, ...effect });
        continue;
      }
      target[effect.property] = effect.after;
    }
  }

  function nextId(edits) {
    let max = 0;
    for (const edit of edits) {
      const match = /^E-([0-9]+)$/.exec(String(edit.id));
      if (match) max = Math.max(max, Number(match[1]));
    }
    return `E-${String(max + 1).padStart(4, "0")}`;
  }

  function validate(kernel, spec, params, today) {
    const declared = spec.parameters;
    const unknown = Object.keys(params).filter(k => !has(declared, k)).sort();
    if (unknown.length) throw new ActionError(`Unknown parameter: ${unknown.join(", ")}.`);

    const values = {};
    const refs = {};
    for (const [name, rule] of Object.entries(declared)) {
      const label = has(rule, "displayName") ? rule.displayName : name;
      let raw = own(params, name);
      if (isString(raw)) raw = raw.trim();
      if (raw === undefined || raw === null || raw === "") {
        if (rule.required) throw new ActionError(`${label} is required.`, name);
        values[name] = null;
        continue;
      }

      const kind = rule.type;
      if (kind === "objectRef") {
        if (!isString(raw)) throw new ActionError(`${label} must be an id.`, name);
        const target = kernel.get(rule.objectType, raw);
        if (target === undefined) throw new ActionError(`There is no ${rule.objectType} with id '${raw}'.`, name);
        refs[name] = [rule.objectType, raw, target];
        values[name] = raw;
      } else if (kind === "string" || kind === "text") {
        if (!isString(raw)) throw new ActionError(`${label} must be text.`, name);
        if (codePoints(raw) < (rule.minLength ?? 0)) {
          throw new ActionError(`${label} needs at least ${rule.minLength} characters.`, name);
        }
        if (codePoints(raw) > (rule.maxLength ?? 1000)) throw new ActionError(`${label} is too long.`, name);
        values[name] = raw;
      } else if (kind === "enum") {
        if (!isString(raw) || !rule.options.includes(raw)) {
          throw new ActionError(`${label} must be one of: ${rule.options.join(", ")}.`, name);
        }
        values[name] = raw;
      } else if (kind === "date") {
        const day = isString(raw) ? parseDate(raw) : null;
        if (day === null) throw new ActionError(`${label} must be a date (YYYY-MM-DD).`, name);
        if (rule.notBefore === "today" && day < today) throw new ActionError(`${label} can't be in the past.`, name);
        if (rule.notAfter === "today" && day > today) throw new ActionError(`${label} can't be in the future.`, name);
        values[name] = day;
      } else {
        throw new ActionError(`Unsupported parameter type '${kind}'.`);
      }
    }
    return { values, refs };
  }

  function check(kernel, criterion, refs) {
    const [objectType, primaryKey, target] = refs[criterion.object];
    if (criterion.type === "propertyIn") {
      if (!criterion.values.includes(target[criterion.property] ?? null)) throw new ActionError(criterion.message);
    } else if (criterion.type === "noLinked") {
      const blocking = kernel.linked(objectType, primaryKey, criterion.link)
        .filter(o => Object.entries(criterion.where).every(([prop, allowed]) => allowed.includes(o[prop] ?? null)));
      if (blocking.length) {
        const linkedType = kernel.ontology.linkTypes[criterion.link].to;
        const pkProp = kernel.ontology.objectTypes[linkedType].primaryKey;
        const ids = blocking.map(o => o[pkProp]).sort();
        const shown = ids.slice(0, 3).join(", ") + (ids.length > 3 ? ` and ${ids.length - 3} more` : "");
        throw new ActionError(`${criterion.message}: ${shown}.`);
      }
    } else {
      throw new ActionError(`Unsupported submission criterion '${criterion.type}'.`);
    }
  }

  function evaluate(expression, values, now) {
    if (isObject(expression)) {
      if (has(expression, "param")) return values[expression.param];
      if (expression.now) return now;
      throw new ActionError(`Unsupported value expression ${JSON.stringify(expression)}.`);
    }
    return expression;
  }

  function plan(ontology, spec, values, refs, now) {
    const effects = [];
    for (const edit of spec.edits) {
      if (edit.type !== "modify") throw new ActionError(`Unsupported edit type '${edit.type}'.`);
      const [objectType, primaryKey, target] = refs[edit.object];
      const properties = ontology.objectTypes[objectType].properties;
      for (const [prop, expression] of Object.entries(edit.set)) {
        if (!(own(properties, prop) || {}).editable) {
          throw new ActionError(`${objectType}.${prop} is not editable by actions.`);
        }
        const after = evaluate(expression, values, now);
        const before = target[prop] ?? null;
        if (before !== after) effects.push({ objectType, primaryKey, property: prop, before, after });
      }
    }
    return effects;
  }

  function createKernel(ontology, objects, savedEdits = []) {
    const edits = [];
    const orphaned = [];
    let skipped = 0;
    for (const edit of savedEdits) {
      if (!isEdit(edit)) {
        skipped += 1;
        continue;
      }
      applyEdit(objects, edit, orphaned);
      edits.push(edit);
    }

    const kernel = {
      ontology,
      objects,
      edits,
      orphaned,
      generatedAt: null,
      loadWarning: skipped ? `${skipped} saved action(s) couldn't be read and were skipped.` : null,

      get(objectType, primaryKey) {
        return own(own(objects, objectType), primaryKey);
      },

      linked(objectType, primaryKey, linkName) {
        const link = own(ontology.linkTypes, linkName);
        if (!link || link.from !== objectType) throw new Error(`Link ${linkName} doesn't start at ${objectType}`);
        return Object.values(own(objects, link.to) || {}).filter(o => o[link.foreignKey] === primaryKey);
      },

      snapshot(today) {
        const copy = {};
        for (const [kind, rows] of Object.entries(objects)) {
          copy[kind] = {};
          for (const [pk, obj] of Object.entries(rows)) copy[kind][pk] = { ...obj };
        }
        return {
          objects: copy,
          generatedAt: kernel.generatedAt,
          today,
          editCount: edits.length,
          orphanedEffects: orphaned.length,
          loadWarning: kernel.loadWarning,
        };
      },

      submit(actionName, params, actor, today, now = nowIso()) {
        if (params === null || params === undefined) params = {};
        if (!isObject(params)) throw new ActionError("params must be an object.");
        if (!isObject(actor) || !isString(actor.role)) throw new ActionError("actor must be an object with a role.");

        const spec = own(ontology.actionTypes, actionName);
        if (!spec) throw new ActionError(`Unknown action '${actionName}'.`);
        if (!spec.submitters.includes(actor.role)) {
          const allowed = spec.submitters.map(r => ontology.roles[r]).join(" or ");
          throw new ActionError(`Only a ${allowed.toLowerCase()} can ${spec.displayName.toLowerCase()}.`);
        }

        const { values, refs } = validate(kernel, spec, params, today);
        for (const criterion of spec.submissionCriteria || []) check(kernel, criterion, refs);
        const effects = plan(ontology, spec, values, refs, now);

        const name = actor.name;
        const edit = {
          id: nextId(edits),
          at: now,
          actor: { name: isString(name) && name.trim() ? name : "unknown", role: actor.role },
          action: actionName,
          params: values,
          effects,
        };
        applyEdit(objects, edit, orphaned);
        edits.push(edit);
        return edit;
      },
    };
    return kernel;
  }

  // Port of src/views.py: derived properties and headline numbers.
  function fleetView(snapshot) {
    const { today, objects } = snapshot;
    const vehicles = objects.Vehicle;
    const orders = objects.WorkOrder;

    for (const vehicle of Object.values(vehicles)) vehicle.openCritical = [];
    for (const order of Object.values(orders)) {
      order.overdue = ACTIVE.has(order.status) && order.dueBy < today;
      if (order.priority === "critical" && ACTIVE.has(order.status)) {
        vehicles[order.vehicleId].openCritical.push(order.workOrderId);
      }
    }
    for (const vehicle of Object.values(vehicles)) {
      vehicle.atRisk = vehicle.status === "in_service" && vehicle.openCritical.length > 0;
    }

    const count = (rows, test) => Object.values(rows).filter(test).length;
    return {
      objects,
      kpis: {
        atRisk: count(vehicles, v => v.atRisk),
        grounded: count(vehicles, v => v.status === "grounded"),
        active: count(orders, o => ACTIVE.has(o.status)),
        overdue: count(orders, o => o.overdue),
        closed: count(orders, o => CLOSED.has(o.status)),
        vehicles: Object.keys(vehicles).length,
        workOrders: Object.keys(orders).length,
      },
      meta: {
        generatedAt: snapshot.generatedAt,
        today,
        editCount: snapshot.editCount,
        orphanedEffects: snapshot.orphanedEffects,
        loadWarning: snapshot.loadWarning,
      },
    };
  }

  const api = { ActionError, createKernel, fleetView, nowIso };
  if (typeof module === "object" && module.exports) module.exports = api;
  else root.RecallKernel = api;
})(typeof globalThis !== "undefined" ? globalThis : this);

File: src/kernel.py
"""The ontology kernel: typed objects, links, and actions as the only write path.

A small, generic interpreter for ontology/ontology.json. It knows nothing
about vehicles or recalls. Everything domain-specific (object types, links,
who may submit each action, what it validates, what it edits) is declared in
the JSON, so adding an action that uses the supported parameter, criterion
and edit types is a data change, not a code change.

State has two layers:
  state/objects.json   rebuilt by the pipeline from source data
  state/edits.jsonl    append-only log of every action a person submitted

Edits are replayed over the pipeline output on load, so re-running the
pipeline never erases a human decision. Each edit records who acted, the
parameters, and the before and after value of every property it touched.
The kernel assumes it is the only process writing the log.

Reading the log fails loudly, with one exception. A torn final line (the
process died mid-append) is moved to edits.jsonl.torn and the rest loads.
Corruption anywhere else stops the kernel and names the line: deciding which
human decisions to keep is not the kernel's call.
"""

import json
import os
import pathlib
import re
import threading
from datetime import date, datetime, timezone

ROOT = pathlib.Path(__file__).resolve().parents[1]


class ActionError(Exception):
    """A submission was rejected. `field` names the offending parameter, if any."""

    def __init__(self, message, field=None):
        super().__init__(message)
        self.message = message
        self.field = field


class LogCorrupted(Exception):
    """The edit log has a damaged line that is not a torn final append."""


def _now():
    return datetime.now(timezone.utc).isoformat(timespec="seconds")


def _signature(path):
    """Cheap change detection: (inode, size, mtime), or None if the file is missing."""
    try:
        stat = path.stat()
    except FileNotFoundError:
        return None
    return (stat.st_ino, stat.st_size, stat.st_mtime_ns)


def _write_atomic(path, text):
    temp = path.with_name(f".{path.name}.tmp-{os.getpid()}")
    temp.write_text(text)
    os.replace(temp, path)


class Kernel:
    def __init__(self,
                 ontology_path=ROOT / "ontology" / "ontology.json",
                 objects_path=ROOT / "state" / "objects.json",
                 edits_path=ROOT / "state" / "edits.jsonl",
                 today=None):
        self.ontology_path = pathlib.Path(ontology_path)
        self.objects_path = pathlib.Path(objects_path)
        self.edits_path = pathlib.Path(edits_path)
        self._today = today
        self._lock = threading.RLock()
        self.load_warning = None
        self.reload()

    def today(self):
        return self._today or date.today()

    # -- loading ------------------------------------------------------------

    def reload(self):
        with self._lock:
            # Signatures are taken before reading, so a write that lands mid-read
            # shows up as a change on the next check instead of being missed.
            objects_signature = _signature(self.objects_path)
            edits_signature = _signature(self.edits_path)

            ontology = json.loads(self.ontology_path.read_text())
            base = json.loads(self.objects_path.read_text())
            objects = base["objects"]
            edits, orphaned, warning = self._read_log(objects)
            if warning:
                edits_signature = _signature(self.edits_path)

            # Swap in the new state only once everything above has succeeded.
            # This also clears any warning left over from an earlier failed reload.
            self.ontology = ontology
            self.objects = objects
            self.generated_at = base.get("generatedAt")
            self.edits = edits
            self.orphaned = orphaned
            self.load_warning = warning
            self._signatures = (objects_signature, edits_signature)

    @staticmethod
    def _parse_edit(line):
        """Parse one log line, or raise ValueError if it isn't a well-formed edit."""
        edit = json.loads(line)
        if not isinstance(edit, dict) or not isinstance(edit.get("id"), str) or not isinstance(edit.get("effects"), list):
            raise ValueError("not an edit")
        for effect in edit["effects"]:
            if not (isinstance(effect, dict) and "after" in effect
                    and all(isinstance(effect.get(k), str) for k in ("objectType", "primaryKey", "property"))):
                raise ValueError("malformed effect")
        return edit

    def _read_log(self, objects):
        """Returns (edits, orphaned effects, warning or None). Applies edits to `objects`."""
        edits, orphaned = [], []
        if not self.edits_path.exists():
            return edits, orphaned, None

        content = self.edits_path.read_text()
        lines = content.split("\n")
        # Every append ends with a newline, so a complete log splits into a
        # trailing empty string. Anything else there is an append that died.
        tail = lines.pop()

        for number, line in enumerate(lines, start=1):
            if not line.strip():
                continue
            try:
                edit = self._parse_edit(line)
            except ValueError:
                raise LogCorrupted(f"{self.edits_path} line {number} is damaged. "
                                   "Fix or remove it by hand; the kernel won't guess.") from None
            self._apply(objects, edit, orphaned)
            edits.append(edit)

        if not tail.strip():
            return edits, orphaned, None

        try:
            edit = self._parse_edit(tail)
        except ValueError:
            with self.edits_path.with_name(self.edits_path.name + ".torn").open("a") as handle:
                handle.write(tail + "\n")
            warning = "A torn final line in the edit log was moved to edits.jsonl.torn."
        else:
            # Complete JSON that only lost its newline: keep it.
            self._apply(objects, edit, orphaned)
            edits.append(edit)
            warning = "The edit log's last line was missing its newline; it was repaired."
        _write_atomic(self.edits_path, "".join(json.dumps(e) + "\n" for e in edits))
        return edits, orphaned, warning

    def refresh_if_rebuilt(self):
        """Pick up a pipeline rebuild, or an edited or deleted log, without a restart."""
        current = (_signature(self.objects_path), _signature(self.edits_path))
        if current == self._signatures:
            return
        try:
            self.reload()
        except (OSError, ValueError, KeyError, TypeError, LogCorrupted) as error:
            # Keep serving the last consistent state rather than half of a new one.
            # The next successful reload clears this warning.
            self.load_warning = f"Reload failed, still serving the previous state: {error}"

    @staticmethod
    def _apply(objects, edit, orphaned):
        for effect in edit["effects"]:
            target = objects.get(effect["objectType"], {}).get(effect["primaryKey"])
            if target is None:
                # The pipeline no longer produces this object. Keep the edit in
                # the log, but surface it rather than silently dropping it.
                orphaned.append({"edit": edit["id"], **effect})
                continue
            target[effect["property"]] = effect["after"]

    # -- reading ------------------------------------------------------------

    def get(self, object_type, primary_key):
        return self.objects.get(object_type, {}).get(primary_key)

    def linked(self, object_type, primary_key, link_name):
        link = self.ontology["linkTypes"][link_name]
        if link["from"] != object_type:
            raise KeyError(f"Link {link_name} starts at {link['from']}, not {object_type}")
        return [o for o in self.objects.get(link["to"], {}).values() if o.get(link["foreignKey"]) == primary_key]

    def snapshot(self):
        """A consistent copy of current state for read models to work from."""
        with self._lock:
            self.refresh_if_rebuilt()
            return {
                "objects": {kind: {pk: dict(obj) for pk, obj in rows.items()} for kind, rows in self.objects.items()},
                "generatedAt": self.generated_at,
                "today": self.today().isoformat(),
                "editCount": len(self.edits),
                "orphanedEffects": len(self.orphaned),
                "loadWarning": self.load_warning,
            }

    def edit_log(self):
        with self._lock:
            self.refresh_if_rebuilt()
            return list(self.edits)

    # -- writing ------------------------------------------------------------

    def submit(self, action_name, params, actor):
        if params is None:
            params = {}
        if not isinstance(params, dict):
            raise ActionError("params must be an object.")
        if not isinstance(actor, dict) or not isinstance(actor.get("role"), str):
            raise ActionError("actor must be an object with a role.")

        with self._lock:
            self.refresh_if_rebuilt()
            spec = self.ontology["actionTypes"].get(action_name)
            if spec is None:
                raise ActionError(f"Unknown action '{action_name}'.")

            if actor["role"] not in spec["submitters"]:
                allowed = " or ".join(self.ontology["roles"][r] for r in spec["submitters"])
                raise ActionError(f"Only a {allowed.lower()} can {spec['displayName'].lower()}.")

            values, refs = self._validate(spec, params)
            for criterion in spec.get("submissionCriteria", []):
                self._check(criterion, refs)
            effects = self._plan(spec, values, refs)

            name = actor.get("name")
            edit = {
                "id": self._next_id(),
                "at": _now(),
                "actor": {"name": name if isinstance(name, str) and name.strip() else "unknown", "role": actor["role"]},
                "action": action_name,
                "params": values,
                "effects": effects,
            }
            self._append(edit)
            self._apply(self.objects, edit, self.orphaned)
            self.edits.append(edit)
            return edit

    def _next_id(self):
        numbers = [int(m.group(1)) for e in self.edits if (m := re.fullmatch(r"E-(\d+)", str(e.get("id"))))]
        return f"E-{max(numbers, default=0) + 1:04d}"

    def _append(self, edit):
        self.edits_path.parent.mkdir(parents=True, exist_ok=True)
        with self.edits_path.open("a") as handle:
            handle.write(json.dumps(edit) + "\n")
            handle.flush()
            os.fsync(handle.fileno())
        # Our own append is not an outside change, so don't trigger a reload for it.
        self._signatures = (self._signatures[0], _signature(self.edits_path))

    def _validate(self, spec, params):
        declared = spec["parameters"]
        unknown = sorted(set(params) - set(declared))
        if unknown:
            raise ActionError(f"Unknown parameter: {', '.join(unknown)}.")

        values, refs = {}, {}
        for name, rule in declared.items():
            label = rule.get("displayName", name)
            raw = params.get(name)
            if isinstance(raw, str):
                raw = raw.strip()
            if raw is None or raw == "":
                if rule.get("required"):
                    raise ActionError(f"{label} is required.", name)
                values[name] = None
                continue

            kind = rule["type"]
            if kind == "objectRef":
                if not isinstance(raw, str):
                    raise ActionError(f"{label} must be an id.", name)
                target = self.get(rule["objectType"], raw)
                if target is None:
                    raise ActionError(f"There is no {rule['objectType']} with id '{raw}'.", name)
                refs[name] = (rule["objectType"], raw, target)
                values[name] = raw
            elif kind in ("string", "text"):
                if not isinstance(raw, str):
                    raise ActionError(f"{label} must be text.", name)
                if len(raw) < rule.get("minLength", 0):
                    raise ActionError(f"{label} needs at least {rule['minLength']} characters.", name)
                if len(raw) > rule.get("maxLength", 1000):
                    raise ActionError(f"{label} is too long.", name)
                values[name] = raw
            elif kind == "enum":
                if not isinstance(raw, str) or raw not in rule["options"]:
                    raise ActionError(f"{label} must be one of: {', '.join(rule['options'])}.", name)
                values[name] = raw
            elif kind == "date":
                # Exactly YYYY-MM-DD. Newer Pythons' fromisoformat also accepts forms
                # like 20260930, and the browser kernel must agree with this one.
                if not isinstance(raw, str) or not re.fullmatch(r"[0-9]{4}-[0-9]{2}-[0-9]{2}", raw):
                    raise ActionError(f"{label} must be a date (YYYY-MM-DD).", name)
                try:
                    day = date.fromisoformat(raw)
                except ValueError:
                    raise ActionError(f"{label} must be a date (YYYY-MM-DD).", name) from None
                if rule.get("notBefore") == "today" and day < self.today():
                    raise ActionError(f"{label} can't be in the past.", name)
                if rule.get("notAfter") == "today" and day > self.today():
                    raise ActionError(f"{label} can't be in the future.", name)
                values[name] = day.isoformat()
            else:
                raise ActionError(f"Unsupported parameter type '{kind}'.")
        return values, refs

    def _check(self, criterion, refs):
        object_type, primary_key, target = refs[criterion["object"]]
        kind = criterion["type"]
        if kind == "propertyIn":
            if target.get(criterion["property"]) not in criterion["values"]:
                raise ActionError(criterion["message"])
        elif kind == "noLinked":
            blocking = [
                o for o in self.linked(object_type, primary_key, criterion["link"])
                if all(o.get(prop) in allowed for prop, allowed in criterion["where"].items())
            ]
            if blocking:
                linked_type = self.ontology["linkTypes"][criterion["link"]]["to"]
                pk_prop = self.ontology["objectTypes"][linked_type]["primaryKey"]
                ids = sorted(o[pk_prop] for o in blocking)
                shown = ", ".join(ids[:3]) + (f" and {len(ids) - 3} more" if len(ids) > 3 else "")
                raise ActionError(f"{criterion['message']}: {shown}.")
        else:
            raise ActionError(f"Unsupported submission criterion '{kind}'.")

    def _plan(self, spec, values, refs):
        effects = []
        now = _now()
        for edit in spec["edits"]:
            if edit["type"] != "modify":
                raise ActionError(f"Unsupported edit type '{edit['type']}'.")
            object_type, primary_key, target = refs[edit["object"]]
            properties = self.ontology["objectTypes"][object_type]["properties"]
            for prop, expression in edit["set"].items():
                if not properties.get(prop, {}).get("editable"):
                    raise ActionError(f"{object_type}.{prop} is not editable by actions.")
                after = self._evaluate(expression, values, now)
                before = target.get(prop)
                if before != after:
                    effects.append({
                        "objectType": object_type,
                        "primaryKey": primary_key,
                        "property": prop,
                        "before": before,
                        "after": after,
                    })
        return effects

    @staticmethod
    def _evaluate(expression, values, now):
        if isinstance(expression, dict):
            if "param" in expression:
                return values[expression["param"]]
            if expression.get("now"):
                return now
            raise ActionError(f"Unsupported value expression {expression}.")
        return expression
